### PR TITLE
Refactor/edit bot responses

### PR DESCRIPTION
**Pull request by kenrilea**
If the feature is a refactor, please explain why your way is better

### CODE REVIEW

**Comment by kenrilea** [comment]
The bot response popup is working as expected, a few tests need to be written, and there is a bug with the bot response page

- Fix failing tests

NEW TESTS
- should not allow duplicate names when editing a response
- it should not allow duplicate names when creating a response require response names to start with utter_
- it should disable edit-reponse-input if the response is used in a story

TODO
- botResponse.utils.js -> checkResponseEmpty is used to prevent an empty response from being created. currently any text response without content text is discarded along with quick replies that have no buttons. A decision should be made on what qualifies as an empty response and the function modified accordingly
- if a response does not exist but is in a story the response editor for that response in the visual story editor cannot be opened


BUGS
- bot response page, starting with no responses create a response then try to delete it.  It is deleted in the database but the subscription does not update the bot response page

File: botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        if (isNew) {
+            setNewBotResponse(
+                { ...(newBotResponse || botResponse), metadata: updatedMetadata });
+            return;
+        }
+        updateResponse({ ...botResponse, metadata: updatedMetadata }, () => {});
+    };
+
+    const handleChangeKey = async () => {
+        if (!responseKey.match(/^utter_/)) {
```
Comment (znat) [inline botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx]: Shouldn't that be enforced at the schema level?

File: botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                if (err) {
+                    console.log(err);
+                    if (err.message.match(/E11000/)) {
+                        setRenameError('Response names must be unique');
+                        return;
+                    }
+                }
+            });
+        }
+    };
```
Comment (znat) [inline botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx]: seems like there's some duplication in validations

File: botfront/imports/ui/components/templates/botResponse.utils.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    if (Object.keys(content).includes('text') && !Object.keys(content).includes('image') && !Object.keys(content).includes('buttons')) return 'TextPayload';
+    return 'CustomPayload';
+};
+
+export const addContentType = content => ({ ...content, __typename: parseContentType(content) });
+
+export const addResponseLanguage = (response, language) => {
+    const updatedResponse = response;
+    const newValue = {
+        sequence: [{ content: safeDump(defaultTemplate(parseContentType(safeLoad(response.values[0].sequence[0].content)))) }],
```
Comment (kenrilea) [inline botfront/imports/ui/components/templates/botResponse.utils.js]: refactoring this line to increase readability

**Comment by znat** [comment]
I created a new branch from this PR (https://github.com/botfront/botfront/pull/403) that will be merged until all issues here are resolved. In addition to the issues in code there are a few bugs:

1. It is not possible to add a custom response from the story editor
Capture d’écran 2020-01-03 à 15 53 08
2. Creating a custom response in the responses and adding it directly in the markdown overrides the responses created with a new one.
3. Please add a tooltip on the delete button when not deletable "This response cannot be deleted because it is used in a story"
Capture d’écran 2020-01-03 à 15 53 33

File: botfront/imports/api/graphql/botResponses/mongo/botResponses.js [hunk 4]
```diff
@@ -41,10 +41,13 @@ export const createResponses = async (projectId, responses) => {
...
-    ...clearTypenameField(newResponse),
-    projectId,
-});
+
+export const createResponse = async (projectId, newResponse) => {
+    return BotResponses.create({
+        ...clearTypenameField(newResponse),
+        projectId,
+    });
+};
```
[the remark this answers is not in the document]
Comment (kenrilea) [reply botfront/imports/api/graphql/botResponses/mongo/botResponses.js]: it's not necessary, reverted.

File: botfront/imports/ui/components/templates/MetadataForm.jsx [hunk 5]
```diff
@@ -108,11 +108,11 @@ function ResponseMetadataForm({
     const validator = (model) => {
         const errors = [...getPageChangeErrors(model)];
 
-        if (model.customCss.enabled && !model.customCss.text && !model.customCssContainer) {
+        if (model.customCss && model.customCss.enabled && !model.customCss.text && !model.customCssContainer) {
             errors.push({ name: 'customCss', message: 'You enabled Custom CSS but you set neither text nor message container properties' });
         }
 
-        if (model.domHighlight.enabled && ((!model.domHighlight.selector || !model.domHighlight.selector.length) || (!model.domHighlight.css || !model.domHighlight.css.length))) {
+        if (model.domHighlight && model.domHighlight.enabled && ((!model.domHighlight.selector || !model.domHighlight.selector.length) || (!model.domHighlight.css || !model.domHighlight.css.length))) {
```
[the remark this answers is not in the document]
Comment (kenrilea) [reply botfront/imports/ui/components/templates/MetadataForm.jsx]: fixed

File: botfront/imports/ui/components/templates/botResponse.utils.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+                sequence: [{ content: safeDump(defaultTemplate(type)) }],
+                lang: language,
+            },
+        ],
+    };
+    return newTemplate;
+};
+
+export const parseContentType = (content) => {
+    if (content.elements || content.attachment || content.custom) return 'CustomPayload';
```
[the remark this answers is not in the document]
Comment (kenrilea) [reply botfront/imports/ui/components/templates/botResponse.utils.js]: fixed

File: botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx (hunk 1 above)
Comment (kenrilea) [reply botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx]: this is now enforced at the schema level

File: botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx (hunk 2 above)
Comment (kenrilea) [reply botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx]: duplication has been removed

**Pull request closed without merging**